fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,573 @@
+use serde::{Deserialize, Serialize};
+use std::fs::{self, File};
+use std::io::{Read, Write};
+use std::time::SystemTime;
+use std::{
+    collections::BTreeMap,
+    path::{Path, PathBuf},
+};
+
+use crate::base_storage::{BaseStorage, SyncStatus};
+use crate::monoid::Monoid;
+// use crate::utils::read_version_2_fs;
+use data_error::{ArklibError, Result};
+
+/*
+Note on `FolderStorage` Versioning:
+
+`FolderStorage` is a basic key-value storage system that persists data to disk.
+where the key is the path of the file inside the directory.
+
+
+In version 2, `FolderStorage` stored data in a plaintext format.
+Starting from version 3, data is stored in JSON format.
+
+For backward compatibility, we provide a helper function `read_version_2_fs` to read version 2 format.
+*/
+const STORAGE_VERSION: i32 = 3;
+
+/// Represents a folder storage system that persists data to disk.
+pub struct FolderStorage<K, V>
+where
+    K: Ord,
+{
+    /// Label for logging
+    label: String,
+    /// Path to the underlying file where data is persisted
+    path: PathBuf,
+    /// `ram_timestamps` can be used to track the last time a file was modified in memory.
+    /// where the key is the path of the file inside the directory.
+    ram_timestamps: BTreeMap<K, SystemTime>,
+    /// `disk_timestamps` can be used to track the last time a file written or read from disk.
+    /// where the key is the path of the file inside the directory.
+    disk_timestamps: BTreeMap<K, SystemTime>,
+    data: FolderStorageData<K, V>,
+}
+
+/// A struct that represents the data stored in a [`FolderStorage`] instance.
+///
+///
+/// This is the data that is serialized and deserialized to and from disk.
+#[derive(Serialize, Deserialize)]
+pub struct FolderStorageData<K, V>
+where
+    K: Ord,
+{
+    version: i32,
```

Comment:
Where should we store the version information? In `FileStorage`, it is serialized in a separate field. For `FolderStorage`, we need a way to track the version to ensure backward compatibility in the future. I have two ideas:

- Embed the storage version in each file's content as a separate field.
- Use a special file in the folder to specify the version.

Reply:
Files storing the values should not be used for this. Examples of values stored in a `FolderStorage` range from JSON properties to binary image data.

Storing the version in a separate file sounds like a better idea to me. But do we really need versioning here?

Reply:
We might manage without versioning information in `FolderStorage`, but if we decide to update to a newer version later (for instance, if binary serialization turns out to be unsuitable), implementing backward compatibility will become more challenging, though not impossible. 

We've already dealt with a similar situation in `FileStorage`.
I don't have a strong opinion on this. If you prefer to omit versioning information, that's fine with me.

Reply:
Agree. We might change the folder structure in future, too. Let's add version file just for the case.

On other hand, versioning of file format probably should be delegated to user side, because we should just write bytes using writer that user passes to us. So, JSONs could be printed as text, or in binary form. Can we implement it in a way that user controls how they want to store JSONs?

Reply:
The proper way to allow the user to pick the serialization format would be to introduce another generic parameter and allow it to pick from a given format, or by making the generic parameter an implementation of Writer. However, I believe this can be a feature for later.

Currently, we can pick JSON as the default format since it is human readable. In fact, later when we extend the implementation for multiple writers for generics we don't need a separate version file because generic parameters can be set with default as JSON.

So for v1, I think not making it generic and using serde_json as the serialization format will be a good approach.